src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -126,14 +139,34 @@ internal partial class Binder
 
             return new BoundConversion(
                 syntax,
-                source,
+                BindToNaturalType(source, diagnostics),
                 conversion,
                 @checked: CheckOverflowAtRuntime,
                 explicitCastInCode: isCast && !wasCompilerGenerated,
                 conversionGroupOpt,
                 constantValueOpt: constantValue,
-                type: destination)
-            { WasCompilerGenerated = wasCompilerGenerated };
+                type: destination) { WasCompilerGenerated = wasCompilerGenerated };
+        }
+
+        /// <summary>
+        /// Rewrite the expressions in the switch expression arms to add a conversion to the destination type.
+        /// </summary>
+        private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression source, TypeSymbol destination, DiagnosticBag diagnostics, bool hasErrors = false)
+        {
+            var builder = ArrayBuilder<BoundSwitchExpressionArm>.GetInstance();
+            foreach (var oldCase in source.SwitchArms)
+            {
+                var oldValue = oldCase.Value;
+                var newValue = GenerateConversionForAssignment(destination, oldValue, diagnostics);
+                var newCase = (oldValue == newValue) ? oldCase :
+                    new BoundSwitchExpressionArm(oldCase.Syntax, oldCase.Locals, oldCase.Pattern, oldCase.WhenClause, newValue, oldCase.Label, oldCase.HasErrors);
```

Comment:
>new BoundSwitchExpressionArm [](start = 20, length = 28)

nit: consider using an `WithValue` method (which can handle the laziness as well)

Reply:
I would if I had another use for it.

---
In reply to: [301761902](https://github.com/dotnet/roslyn/pull/37052#discussion_r301761902) [](ancestors = 301761902)